Update 'time origin' and `workerStart` definitions

igrigorik opened the pull request:
- Clarifies definition of time origin for Window (closes #3)
- Defines time origin for Workers as time of creation of worker context (instead of 'moment of creation')
- Defines workerStart as the difference between worker's time origin and the time origin of incumbent settings object. Note: as a result, workerStart will no longer return negative values.

Mailing list threads:
- https://lists.w3.org/Archives/Public/public-web-perf/2015May/0032.html
- https://lists.w3.org/Archives/Public/public-web-perf/2015May/0013.html

Preview:
- https://rawgit.com/w3c/hr-time/time-origin/index.html#sec-time-origin
- https://rawgit.com/w3c/hr-time/time-origin/index.html#sec-worker-start

---

This update changes the _shipped_ (Chrome, Firefox) definition of time origin for Workers + workerStart. Neither browser has existing telemetry on potential impact of this change, but based on [HTTP Archive analysis of Worker usage](https://lists.w3.org/Archives/Public/public-web-perf/2015Jun/0031.html) _I believe_ the risk is sufficiently small and manageable. 

The benefit is that we provide a more consistent API - e.g. no need to branch based on whether you're in shared or dedicated worker. Also, in case of SharedWorker's it's now possible for each reference to meaningfully translate between timebases of the worker and the context that holds the reference.

bzbarsky commented on index.html:
```diff
@@ -289,20 +292,11 @@ <h3 id="sec-time-origin"><span class="secno">4.2 </span>Time Origin</h3>
-  <li>The <i>time origin</i> must be equal to the time of the start of
-  <a href="http://www.w3.org/TR/html5/browsers.html#navigate">navigation</a> of the current document.</li>
-  <li>The <i>time origin</i> for <a href="http://www.w3.org/TR/workers/#dedicated-workers-and-the-dedicatedworkerglobalscope-interface">dedicated</a> and <a href="http://www.w3.org/TR/workers/#shared-workers-and-the-sharedworkerglobalscope-interface">shared</a> workers must be equal to the time of <a href="http://dev.w3.org/html5/workers/#official-moment-of-creation">creation</a> of the worker co ...
+  <li>If the settings object specified by the incumbent settings object is a <code>Window</code> object, the <i>time origin</i> must be equal to the time of starting the <a href="http://www.w3.org/TR/html5/browsers.html#navigate">navigation</a> responsible for loading the current document, unless a confirmation dialog is displayed during the <a href="http://www.w3.org/TR/html5/browsers.html#promp ...
```
"If the global object specified", no?

bzbarsky commented on index.html:
```diff
@@ -289,20 +292,11 @@ <h3 id="sec-time-origin"><span class="secno">4.2 </span>Time Origin</h3>
-  <a href="http://www.w3.org/TR/html5/browsers.html#navigate">navigation</a> of the current document.</li>
-  <li>The <i>time origin</i> for <a href="http://www.w3.org/TR/workers/#dedicated-workers-and-the-dedicatedworkerglobalscope-interface">dedicated</a> and <a href="http://www.w3.org/TR/workers/#shared-workers-and-the-sharedworkerglobalscope-interface">shared</a> workers must be equal to the time of <a href="http://dev.w3.org/html5/workers/#official-moment-of-creation">creation</a> of the worker co ...
+  <li>If the settings object specified by the incumbent settings object is a <code>Window</code> object, the <i>time origin</i> must be equal to the time of starting the <a href="http://www.w3.org/TR/html5/browsers.html#navigate">navigation</a> responsible for loading the current document, unless a confirmation dialog is displayed during the <a href="http://www.w3.org/TR/html5/browsers.html#promp ...
+  <li>If the settings object specified by the incumbent settings object is a <code>WorkerGlobalScope</code> object, the <i>time origin</i> must be equal to the time <a href="http://dev.w3.org/html5/workers/#set-up-a-worker-script-settings-object">when the worker script settings object is set up</a>.</li>
```
Again, "if the global object".

bzbarsky commented on index.html:
```diff
@@ -359,16 +353,7 @@ <h3 id="sec-worker-start"><span class="secno">4.5 </span>The <code>workerStart</
-  <i><a href="#time-origin">time origin</a></i>.
-</p>
-</div>
+<p>The <dfn id="dom-worker-start"><code>workerStart</code></dfn> attribute MUST return a <a href="#sec-DOMHighResTimeStamp"><code>DOMHighResTimeStamp</code></a> representing the difference between the worker's <i><a href="#time-origin">time origin</a></i> and the <i><a href="#time-origin">time origin</a></i> of the incumbent settings object. If the incumbent settings object's time origin is undef ...
```
Difference in which direction?  Worth spelling it out, probably, esp. since this difference can still be negative in the Shared/ServiceWorker cases.  I'm not sure I have good suggested wording here.  Maybe "the result of subtracting the time origin of the incumbent settings object from the time origin of the worker's WorkerGlobalScope"?

toddreifsteck commented:
Ensuring @digitec sees this. Would like to ensure we have his input.

igrigorik commented:
@bzbarsky thanks, updated, ptal.

Re, workerStart "since the difference can be negative": we've defined worker's time origin as relative to when the reference to the sharedworker is created, as such, I don't think that's true?

toddreifsteck commented:
Justin and I went over this and came up with a scenario we aren't sure this solves.... If you create the Worker from a main document then access it from an iFrame, can all 3 reason about each other? Do we expose sufficient information with this update for that?

bzbarsky commented:
> we've defined worker's time origin as relative to when the reference to the sharedworker is created

The time origin is being defined as when the WorkerGlobalScope is created, no?  This can be before the page using the SharedWorker was loaded, so the workerStart can be negative.

@toddreifsteck This update provides sufficient information for both the main document and the iframe to reason about the worker.  In particular, the main document and the iframe see different workerStart values for the same worker, so each one can convert timestamps in the worker's timeline into its own timeline.  They can also compute the difference between their time origins by communicating to each other what workerStart values they're seeing.

The main document and the iframe can somewhat reason about each other even without the worker by comparing navigationStart times, I think.

igrigorik commented:
Based on the [discussion on the mailing list](https://lists.w3.org/Archives/Public/public-web-perf/2015Jun/0092.html), the new proposal is to replace workerStart with:

```
partial interface Performance {
  DOMHighResTimeStamp translateTime(DOMHighResTimeStamp time, (Window or Worker) timeSource);
};
```
> The `translateTime` method MUST return a DOMHighResTimeStamp that is the result of adding the provided `time`, and the difference of the _time origin_ of the global of the performance object that is the `this` value for the `translateTime` call from the _time origin_ of the provided `timeSource`.

_i.e. time + ((time origin of timeSource.performance) - (time origin of this.performance))_

--- 

>> we've defined worker's time origin as relative to when the reference to the sharedworker is created

> The time origin is being defined as when the WorkerGlobalScope is created, no? This can be before the page using the SharedWorker was loaded, so the workerStart can be negative.

Right now the proposal is to make it: _"If the global object specified by the incumbent settings object is a WorkerGlobalScope object, the time origin must be equal to the time [when the worker script settings object is set up](http://dev.w3.org/html5/workers/#set-up-a-worker-script-settings-object)."_

Looking at http://dev.w3.org/html5/workers/#shared-workers-and-the-sharedworker-interface, that's step 9, and I think it should capture the time when SharedWorker reference is created.. no? At least, that was the intent.

bzbarsky commented:
> and the difference of

I think you're looking for:

    and the difference between the time origin of the provided `timeSource` and the
    time origin of the global of the ... [etc]

And then the informative equation to make it clear.

> and I think it should capture the time when SharedWorker reference is created

When the first-ever SharedWorker reference for that SharedWorkerGlobalScope is created, more precisely.  Or at least I hope that's what you mean.

Unless you really mean step 9 (or rather step 8.9?) always, in which case I have no idea what that would mean when the algorithm returns in step 8.7.9.

igrigorik commented:
> And then the informative equation to make it clear.

Fair enough. I was trying to spell it out in the definition, but perhaps that's a better route.

Added to spec and updated the example in intro:
- https://github.com/w3c/hr-time/commit/39aaf81e1ea206fdf0fde4f49604ddd64d4f5b13
- https://rawgit.com/w3c/hr-time/time-origin/index.html#methods
- https://rawgit.com/w3c/hr-time/time-origin/index.html#examples

How does that look? Note that I'm deprecating `workerStart`.  

> When the first-ever SharedWorker reference for that SharedWorkerGlobalScope is created, more precisely. Or at least I hope that's what you mean.

Hmm, so with translateTime in place (and removing workerStart), I think we've addressed [@sicking's concerns](https://lists.w3.org/Archives/Public/public-web-perf/2015Jun/0008.html). With that in mind, yeah, I guess we want time origin to be the time when the worker is created.. [Official moment of creation](http://dev.w3.org/html5/workers/#official-moment-of-creation)? If not, then we need a hook that makes this clear(er).

toddreifsteck commented:
I think it may be helpful to improve the translateTime example to assign performance.now() on both a Worker and within Window to well named variables and then to use translateTime to clearly demonstrate how to make a timeline out of those values. I believe the current example requires a bit more understanding than ideal for a reader new to this spec.

igrigorik commented:
How about...

```javascript
// Shared worker script (worker.js)
onconnect = function(e) {
  var port = e.ports[0];
  port.onmessage = function(e) {
    // Time execution in worker
    var task_start = performance.now();
    result = runSomeWorkerTask();
    var task_end = performance.now();

    port.postMessage({
       'task': 'Some worker task',
       'start_time': task_start,
       'end_time': task_end,
       'result': result
    });
  }
}
```

```javascript
// Timing tasks in a dedicated or shared worker 
var worker = new SharedWorker('worker.js');
worker.port.onmessage = function (event) {
  var msg = event.data;

  // translate timestamps into document's time origin
  msg.start_time = performance.translateTime(msg.start_time, worker);
  msg.end_time = performance.translateTime(msg.end_time, worker);

  // plot the results on document's timeline
  plotEventOnTimeline(msg);
}

// Timing tasks in the document
var task_start = performance.now();
result = runSomeWorkerTask();
var task_end = performance.now();

plotEventOnTimeline({
  'task': 'Some document task',
  'start_time': task_start,
  'end_time': task_end,
  'result': result
});
```

bzbarsky commented:
> Official moment of creation?

Oh, we have such a thing already. Excellent.  That would make some sense, yes.

DigiTec commented:
This is looking really good Some nits:
`DOMHighResTimeStamp translateTime (DOMHighResTimeStamp time, (Windor or Worker) timeSource);`

Window is misspelled.

We should add SharedWorker to the or set or we should clarify if a mix-in interface (AbstractWorker) can be used in the or group and switch to that instead of Worker. I think, though I'm very uncertain, that there may be other examples of a mixin used as a type discriminator though. Specifically EventTarget was a mixin for a while and we used it as a real type.

bzbarsky commented:
Using a mixin there would be a bit weird, actually.  We should just spell out Worker or SharedWorker or ServiceWorker.

igrigorik commented:
@bzbarsky @toddreifsteck @DigiTec updated in 2961b4a, see preview: https://rawgit.com/w3c/hr-time/time-origin/index.html

toddreifsteck commented:
LGTM

From: Ilya Grigorik [mailto:notifications@github.com]
Sent: Monday, June 22, 2015 1:42 PM
To: w3c/hr-time
Cc: Todd Reifsteck
Subject: Re: [hr-time] Update 'time origin' and `workerStart` definitions (#6)


@bzbarsky<https://github.com/bzbarsky> @toddreifsteck<https://github.com/toddreifsteck> @DigiTec<https://github.com/DigiTec> updated in 2961b4a<https://github.com/w3c/hr-time/commit/2961b4a821bcda2cae17a62b7053a0e367ac0a54>, see preview: https://rawgit.com/w3c/hr-time/time-origin/index.html

—
Reply to this email directly or view it on GitHub<https://github.com/w3c/hr-time/pull/6#issuecomment-114256238>.

bzbarsky commented:
That seems pretty reasonable.  The one problem is that the link to "time origin" is to https://rawgit.com/w3c/hr-time/time-origin/index.html#time-origin which is defined in terms of incumbent settings objects, but we actually want a definition of time origin for globals.  I suggest that section be rewritten to specify the definition of the time origin of a global.

igrigorik commented:
@bzbarsky if I just strike "specified by the incumbent settings" in both clauses, would that cover it?

bzbarsky commented:
I think so, yes.

igrigorik commented:
@bzbarsky done.

@toddreifsteck if what we have here looks good, please feel free to hit the merge button :)
